### PR TITLE
Use RDatasetSpec in distrdf task creation

### LINKED ISSUE #10872: [DF] Wrong entries are loaded from friend trees with distributed RDF

### PR DESCRIPTION
**Pull request by vepadulano**
Fixes #10872

### CODE REVIEW

File: bindings/experimental/distrdf/python/DistRDF/HeadNode.py [hunk 1]
```diff
@@ -358,15 +358,18 @@ def __init__(self, backend: BaseBackend, npartitions: Optional[int], *args):
         super().__init__(backend, npartitions)
 
         self.defaultbranches = None
-        self.friendinfo = None
+        # Information about friend trees. It is not None only if the TTree
+        # constructor overload is used and if there is at least one friend.
```
Comment (eguiraud) [inline bindings/experimental/distrdf/python/DistRDF/HeadNode.py:362]: this comment will go out of date fast: "...only if the TTree constructor _or the RDatasetSpec constructor_"

File: bindings/experimental/distrdf/python/DistRDF/HeadNode.py [hunk 2]
```diff
@@ -358,15 +358,18 @@ def __init__(self, backend: BaseBackend, npartitions: Optional[int], *args):
...
+        self.friendinfo: Optional[ROOT.Internal.TreeUtils.RFriendInfo] = None
 
         # Retrieve the TTree/TChain that will be processed
         if isinstance(args[0], ROOT.TTree):
             # RDataFrame(tree, defaultBranches = {})
             self.tree = args[0]
             # Retrieve information about friend trees when user passes a TTree
             # or TChain object.
-            self.friendinfo = ROOT.Internal.TreeUtils.GetFriendInfo(args[0])
+            _fi = ROOT.Internal.TreeUtils.GetFriendInfo(args[0])
```
Comment (eguiraud) [inline bindings/experimental/distrdf/python/DistRDF/HeadNode.py:371]: the leading underscore confused me for a sec there :thinking: local variables are by construction internal/private, why does this one deserve a `_` and other local variables in other methods do not?

File: bindings/experimental/distrdf/python/DistRDF/HeadNode.py [hunk 3]
```diff
@@ -528,24 +467,21 @@ def build_rdf_from_range(current_range: Ranges.TreeRangePerc) -> TaskObjects:
...
                 return TaskObjects(None, entries_in_trees)
 
-            # Create RDataFrame object for this task
-            if defaultbranches is not None:
-                rdf = ROOT.RDataFrame(chain, defaultbranches)
-            else:
-                rdf = ROOT.RDataFrame(chain)
+            treeandfilenameglobs = [
+                (treename, filename) for treename, filename in zip(clustered_range.treenames, clustered_range.filenames)
+            ]
```
Comment (eguiraud) [inline bindings/experimental/distrdf/python/DistRDF/HeadNode.py:478]: can this be just

```python
treeandfilenameglobs = zip(clustered_range.treenames, clustered_range.filenames)
```
?

**Review by eguiraud — COMMENTED** [review]
reviewed all changes except those in Ranges.py, which as very large and got me stumped for a while until I had to move on. the rest looks good to me.

the commit message should probably mention #10872 explicitly.

File: bindings/experimental/distrdf/python/DistRDF/HeadNode.py (hunk 2 above)
Comment (vepadulano) [reply bindings/experimental/distrdf/python/DistRDF/HeadNode.py:371]: You are right, my brain decided to promote this to a private variable for some reason :)

File: bindings/experimental/distrdf/python/DistRDF/HeadNode.py (hunk 3 above)
Comment (vepadulano) [reply bindings/experimental/distrdf/python/DistRDF/HeadNode.py:478]: I wonder if cppyy can convert the zip iterator to an `std::vector`, I will try

Comment (vepadulano) [reply bindings/experimental/distrdf/python/DistRDF/HeadNode.py:478]: Interestingly, it seems to work. I will modify the code accordingly

File: bindings/experimental/distrdf/python/DistRDF/Ranges.py [hunk 4]
```diff
@@ -234,11 +229,89 @@ def get_percentage_ranges(treenames: List[str], filenames: List[str], npartition
...
+        ]
+    else:
+        # With the indexes created above, we can partition the lists of names of
+        # files and trees. Each task will get a number of trees dictated by the
+        # starting index (inclusive) and the ending index (exclusive)
+        tasktreenames = [treenames[s:e] for s, e in zip(start_sample_idxs, end_sample_idxs)]
+        taskfilenames = [filenames[s:e] for s, e in zip(start_sample_idxs, end_sample_idxs)]
+        # In each task, the starting file index is 0 and the ending file index
+        # is equal to the number of files assigned to that particular task
+        first_file_idxs = [0] * npartitions
```
Comment (vepadulano) [inline bindings/experimental/distrdf/python/DistRDF/Ranges.py:248]: This list can actually be avoided, just passing `0` as starting index for all `TreeRangePerc` objects created below

File: bindings/experimental/distrdf/python/DistRDF/Ranges.py [hunk 5]
```diff
@@ -234,11 +229,89 @@ def get_percentage_ranges(treenames: List[str], filenames: List[str], npartition
...
+        last_file_idxs = [len(fns) for fns in taskfilenames]
```
Comment (vepadulano) [inline bindings/experimental/distrdf/python/DistRDF/Ranges.py:249]: This list can be avoided by calling `len(taskfilenames[rangeid])` in the creation of each `TreeRangePerc` below

**Review by etejedor — APPROVED** [review]
Very nice! Thanks!

**Pull request merged**